### PR TITLE
add userSharingScope field when exporting to Synapse table

### PR DESCRIPTION
**Pull request by zaclinsage**

### CODE REVIEW

**Comment by zaclinsage** [comment]
test done: 
1. all unit tests in package;
2. manual test: upload test data through Bridge Server and send SQS msg to exporter; then verify if new column was added;
3. mvn clean test / mvn clean verify;
4. add more tests: unit tests for TransferMethod, serialize/de-serialize test for ColumnDefinition, test cases for no common column list table

File: src/test/java/org/sagebionetworks/bridge/exporter/handler/HealthDataExportHandlerTest.java [hunk 1]
```diff
@@ -42,7 +47,14 @@
 import org.sagebionetworks.bridge.rest.model.UploadFieldDefinition;
 import org.sagebionetworks.bridge.rest.model.UploadFieldType;
 
-public class HealthDataExportHandlerTest {
+import javax.annotation.Resource;
+
+@ContextConfiguration("classpath:test-context.xml")
+public class HealthDataExportHandlerTest extends AbstractTestNGSpringContextTests {
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/handler/HealthDataExportHandlerTest.java]: Strongly recommend against introducing Spring into unit tests, especially mock tests that don't hit real dependencies. It adds unnecessary overhead and complexity to our tests, makes them more difficult to manage, and you can do this testing without using Spring anyway. This also means you're essentially testing the same thing in 4 different places. Also, if we ever change the configs, we'll need to change the tests in 4 different places, which is exactly what we wanted to avoid doing with this change.

Here and in SynapseExportHandlerNewTableTest, SynapseExportHandlerTest, and SynapseExportHandlerUpdateTableTest.

File: pom.xml [hunk 2]
```diff
@@ -97,6 +97,11 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-logging</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-test</artifactId>
+            <version>4.0.5.RELEASE</version>
+        </dependency>
```
Comment (DwayneJengSage) [inline pom.xml]: We don't want to introduce Spring into our unit tests. Please remove.

File: src/main/java/org/sagebionetworks/bridge/exporter/config/SpringConfig.java [hunk 3]
```diff
@@ -194,4 +200,15 @@ public SynapseClient synapseClient() {
     public ExecutorService workerExecutorService() {
         return Executors.newFixedThreadPool(bridgeConfig().getInt("threadpool.worker.count"));
     }
+
+    @Bean(name = "synapseColumnDefinitions")
+    public List<ColumnDefinition> synapseColumnDefinitions() throws IOException {
+        final ClassLoader classLoader = getClass().getClassLoader();
+        final File file = new File(classLoader.getResource("ColumnDefinition.json").getFile());
+        final ObjectMapper mapper = new ObjectMapper();
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/config/SpringConfig.java]: Minor: Instead of instantiating your own ObjectMapper, you can use DefaultObjectMapper.INSTANCE from bridge-base.

File: src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java [hunk 4]
```diff
@@ -47,8 +50,14 @@
     private static final Logger LOG = LoggerFactory.getLogger(SynapseExportHandler.class);
 
     // Package-scoped to be available to unit tests.
-    static final List<ColumnModel> COMMON_COLUMN_LIST;
-    static {
+    List<ColumnModel> COMMON_COLUMN_LIST;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java]: Minor: Should be commonColumnList, not COMMON_COLUMN_LIST.

File: src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java [hunk 5]
```diff
@@ -47,8 +50,14 @@
...
+
+    private List<ColumnDefinition> COLUMN_DEFINITION;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java]: Similarly here.

File: src/main/java/org/sagebionetworks/bridge/exporter/config/SpringConfig.java [hunk 6]
```diff
@@ -194,4 +200,15 @@ public SynapseClient synapseClient() {
...
+
+        List<ColumnDefinition> value = mapper.readValue(file, new TypeReference<List<ColumnDefinition>>(){});
+
+        return value;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/config/SpringConfig.java]: Recommend copying the result into an ImmutableList. This prevents consuming classes from accidentally modifying the bean.

Also, our manual test case should have multiple tables to make sure we're handling this correctly.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/TransferMethod.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        public String transfer(String ddbName, Item record, ExportTask task) {
+            return String.valueOf(record.getLong(ddbName));
+        }
+    },
+    EXPORTERDATE {
+        @Override
+        public String transfer(String ddbName, Item record, ExportTask task) {
+            return task.getExporterDate().toString();
+        }
+    };
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/TransferMethod.java:57]: I don't think ExporterDate is useful. This is specific to one column, and the ExportTask is just exporter state and is unlikely to have new fields that need exporting. We should remove this.

File: src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java [hunk 8]
```diff
@@ -57,36 +66,6 @@
...
-        dataGroupsColumn.setColumnType(ColumnType.STRING);
-        dataGroupsColumn.setMaximumSize(100L);
-        columnListBuilder.add(dataGroupsColumn);
-
-        // NOTE: ColumnType.DATE is actually a timestamp. There is no calendar date type.
-        ColumnModel uploadDateColumn = new ColumnModel();
-        uploadDateColumn.setName("uploadDate");
-        uploadDateColumn.setColumnType(ColumnType.STRING);
-        uploadDateColumn.setMaximumSize(10L);
-        columnListBuilder.add(uploadDateColumn);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java:83]: Similarly, don't delete uploadDateColumn here.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    private String name;
+    private ColumnType columnType;
+    private Long maximumSize;
+    private TransferMethod transferMethod;
+    private String ddbName;
+    private Boolean sanitize;
+
+    public ColumnDefinition() {
+
+    }
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java]: Minor: Don't need to explicitly create a constructor if it has no args and no body.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+import org.sagebionetworks.repo.model.table.ColumnType;
+
+public class ColumnDefinition  {
+    private String name;
+    private ColumnType columnType;
+    private Long maximumSize;
+    private TransferMethod transferMethod;
+    private String ddbName;
+    private Boolean sanitize;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java]: Minor: For sanitize, since we treat the null case the same as the false case, we could just use a lower-case boolean and default to false. This will save us null checks elsewhere.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        int result1 = 31 * result + (this.name == null?0:this.name.hashCode());
+        result1 = 31 * result1 + (this.columnType == null?0:this.columnType.hashCode());
+        result1 = 31 * result1 + (this.maximumSize == null?0:this.maximumSize.hashCode());
+        result1 = 31 * result1 + (this.transferMethod == null?0:this.transferMethod.hashCode());
+        result1 = 31 * result1 + (this.ddbName == null?0:this.ddbName.hashCode());
+        result1 = 31 * result1 + (this.sanitize == null?0:this.sanitize.hashCode());
+        return result1;
+    }
+
+    public boolean equals(Object obj) {
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java]: Do we use .equals() on ColumnDefinitions anywhere, or use it in a hash map? If not, then we don't need hashCode() and equals().

If we *do* need hashCode() and equals(), we need to test it. I recommend using EqualsVerifier. See BridgeExporterRequestTest for example.

Also, keep in mind that mutable objects don't work as hash keys. If you need to use this as a hash key to a hash map, strongly recommend making this class immutable and use the builder pattern instead.

Lastly, you can use Eclipse or IntellJ to auto-generate hashCode() and equals() to make them less wordy. Recommend using either Guava Objects or Java 8 Objects to generate these.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+                }
+            } else if(!this.sanitize.equals(other.sanitize)) {
+                return false;
+            }
+
+            return true;
+        }
+    }
+
+    public String toString() {
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java]: Similarly, this can be auto-generated. Also, recommend having a test with all null values to verify that this doesn't throw an NPE if we call toString().

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
+package org.sagebionetworks.bridge.exporter.synapse;
+
+import org.sagebionetworks.repo.model.table.ColumnType;
+
+public class ColumnDefinition  {
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java:5]: We need a test that serializes and deserializes to/from JSON to make sure this works the way we expect.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/TransferMethod.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Enum class representing several different transfer method from record value in ddb table to synapse row value.
+ * Also specifying corresponding transfer procedure.
+ */
+public enum TransferMethod {
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/TransferMethod.java:16]: Needs unit tests.

File: src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java [hunk 15]
```diff
@@ -1,15 +1,20 @@
 package org.sagebionetworks.bridge.exporter.util;
 
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java]: Minor: Our convention is to show all our imports instead of collapsing them with *

File: src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java [hunk 16]
```diff
@@ -1,15 +1,20 @@
...
 
 import com.amazonaws.services.dynamodbv2.document.Item;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSetMultimap;
 import org.jsoup.Jsoup;
 import org.jsoup.safety.Whitelist;
+import org.omg.PortableInterceptor.SYSTEM_EXCEPTION;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java]: What is SYSTEM_EXCEPTION?

File: src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java [hunk 17]
```diff
@@ -21,6 +26,8 @@
 public class BridgeExporterUtil {
     private static final Logger LOG = LoggerFactory.getLogger(BridgeExporterUtil.class);
 
+    public static final Joiner STRING_SET_JOINER = Joiner.on(',').useForNull("");
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java:31]: Minor: If we're moving DATA_GROUP_JOINER here, we should delete the original (now unused) one from SynapseExportHandler.

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 18]
```diff
@@ -19,6 +33,12 @@
 import org.sagebionetworks.bridge.schema.UploadSchemaKey;
 
 public class BridgeExporterUtilTest {
+    private List<ColumnModel> TEST_COLUMN_MODELS;
+    private List<ColumnDefinition> TEST_COLUMN_DEFINITIONS;
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: Minor: non-static member variables should be camel-cased (testColumnModles), not all-caps (TEST_COLUMN_MODELS).

Also, it looks like these aren't being used.

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 19]
```diff
@@ -202,4 +222,132 @@ public void shouldConvertFreeformTextToAttachment() {
                 .withRevision(1).build();
         assertFalse(BridgeExporterUtil.shouldConvertFreeformTextToAttachment(other, "other"));
     }
+
+    @Test
+    public void canConvertToColumnList() {
+        final String TEST_COLUMN_MODEL_NAME_1 = "test_column_model_1";
+        final String TEST_COLUMN_MODEL_NAME_2 = "test_column_model_2";
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: Local variables should also be camel-cased.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    private String name;
+    private ColumnType columnType;
+    private Long maximumSize;
+    private TransferMethod transferMethod;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java:8]: Can we combine TransferMethod and ColumnType? It looks like these map one-to-one, and I can't think of any case where they'd be different. The simplest thing to do is for TransferMethod to have a getColumnType() method that returns the appropriate ColumnType.

File: src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java [hunk 21]
```diff
@@ -185,4 +192,44 @@ public static boolean shouldConvertFreeformTextToAttachment(UploadSchemaKey sche
...
+
+        return columnListBuilder.build();
+    }
+
+    public static Map<String, String> getRowValuesFromRecordBasedOnColumnDefinition(final Map<String, String> rowMap, final Item record, final List<ColumnDefinition> columnDefinitions, final String recordId, final ExportTask task) {
+        Map<String, String> retMap = new HashMap<>();
+        retMap.putAll(rowMap);
+
+        for (ColumnDefinition columnDefinition : columnDefinitions) {
+            final String ddbName = columnDefinition.getDdbName();
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java]: Would it be worth adding a feature so that if ddbName is null, we use name for the ddbName? This way, we don't always have to type every name twice in the JSON configs.

File: src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java [hunk 22]
```diff
@@ -185,4 +192,44 @@ public static boolean shouldConvertFreeformTextToAttachment(UploadSchemaKey sche
...
+            columnModel.setMaximumSize(columnDefinition.getMaximumSize());
+            columnListBuilder.add(columnModel);
+        }
+
+        return columnListBuilder.build();
+    }
+
+    public static Map<String, String> getRowValuesFromRecordBasedOnColumnDefinition(final Map<String, String> rowMap, final Item record, final List<ColumnDefinition> columnDefinitions, final String recordId, final ExportTask task) {
+        Map<String, String> retMap = new HashMap<>();
+        retMap.putAll(rowMap);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtil.java]: Instead of intsantiating a map, copying the contents, and returning the new map, maybe we can have this method update the map in-place and have a void return type. This way, we don't have to instantiate more maps then we need, and it makes the code in SynapseExportHelper easier to reason about. (Right now, we take a map, copy it, and then add all those values back to itself.)

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 23]
```diff
@@ -202,4 +222,132 @@ public void shouldConvertFreeformTextToAttachment() {
...
+        assertEquals(BridgeExporterUtil.convertToColumnList(TEST_COLUMN_DEFINITIONS), TEST_COLUMN_LIST);
+    }
+
+    @Test
+    public void canGetRowValuesFromRecordBasedOnColumnDefinition() {
+        final String TEST_STRING_NAME = "test_string";
+        final String TEST_STRING_SET_NAME = "test_string_set";
+        final String TEST_DATE_NAME = "test_date";
+        final String TEST_EXPORTER_DATE_NAME = "test_exporter_date";
+        final String TEST_SANITIZE = "test_sanitize";
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: Similarly here.

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 24]
```diff
@@ -202,4 +222,132 @@ public void shouldConvertFreeformTextToAttachment() {
...
+        ExportTask mockTask = mock(ExportTask.class);
+        when(mockTask.getExporterDate()).thenReturn(TEST_LOCAL_DATE);
+
+        // create mock record
+        Item testRecord = new Item();
+        testRecord.withString(TEST_STRING_NAME, "test_string_value");
+        testRecord.withStringSet(TEST_STRING_SET_NAME, new String[]{"test_string_set_value_1", "test_string_set_value_2"});
+        testRecord.withLong(TEST_DATE_NAME, 1484181511);
+        testRecord.withString(TEST_EXPORTER_DATE_NAME, "test_exporter_date_value");
+        testRecord.with(TEST_SANITIZE, "test_sanitize_value");
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: The test_sanitize_value should have some HTML in it or something to verify we're properly calling the sanitization code.

File: src/main/resources/ColumnDefinition.json [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+    "transferMethod": "STRINGSET",
+    "ddbName": "userDataGroups"
+  },
+  {
+    "name": "uploadDate",
+    "columnType": "STRING",
+    "maximumSize": "10",
+    "transferMethod": "EXPORTERDATE",
+    "ddbName": "uploadDate"
+  },
```
Comment (DwayneJengSage) [inline src/main/resources/ColumnDefinition.json]: Similarly, remove uploadDate from here.

File: src/main/resources/test-context.xml [hunk 26]
```diff
@@ -0,0 +1,8 @@
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns:context="http://www.springframework.org/schema/context"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans
+        http://www.springframework.org/schema/beans/spring-beans.xsd http://www.springframework.org/schema/context http://www.springframework.org/schema/context/spring-context.xsd">
+    <context:annotation-config/>
+    <bean class="org.sagebionetworks.bridge.exporter.config.SpringConfig"/>
+</beans>
```
Comment (DwayneJengSage) [inline src/main/resources/test-context.xml]: Similarly, remove this.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
You should also add a test to SynapseExportHandlerUpdateTableTest with a table that doesn't have the common columns (or only has some of them) and test that your code successfully adds them to the mock TSV.

File: src/main/java/org/sagebionetworks/bridge/exporter/config/SpringConfig.java [hunk 27]
```diff
@@ -194,4 +202,17 @@ public SynapseClient synapseClient() {
...
+
+    @Bean(name = "synapseColumnDefinitions")
+    public List<ColumnDefinition> synapseColumnDefinitions() throws IOException {
+        final ClassLoader classLoader = getClass().getClassLoader();
+        final File file = new File(classLoader.getResource("ColumnDefinition.json").getFile());
+        final ObjectMapper mapper = DefaultObjectMapper.INSTANCE;
+
+        List<ColumnDefinition> value = mapper.readValue(file, new TypeReference<List<ColumnDefinition>>(){});
+        ImmutableList.Builder<ColumnDefinition> columnDefinitionsBuilder = ImmutableList.builder();
+        columnDefinitionsBuilder.addAll(value);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/config/SpringConfig.java]: Minor: You can use ImmutableList.copyOf(value).

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java [hunk 28]
```diff
@@ -0,0 +1,61 @@
+package org.sagebionetworks.bridge.exporter.synapse;
+
+import org.sagebionetworks.repo.model.table.ColumnType;
+
+public class ColumnDefinition  {
+    private String name;
+    private ColumnType columnType;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/ColumnDefinition.java]: Please remove columnType, as this no longer does anything.

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 29]
```diff
@@ -202,4 +217,123 @@ public void shouldConvertFreeformTextToAttachment() {
...
+        testDefinition2.setMaximumSize(48L);
+        columnDefinitionBuilder.add(testDefinition2);
+
+        testColumnDefinitions = columnDefinitionBuilder.build();
+
+        assertEquals(BridgeExporterUtil.convertToColumnList(testColumnDefinitions), testColumnList);
+    }
+
+    @Test
+    public void canGetRowValuesFromRecordBasedOnColumnDefinition() {
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: We need a test case with name but not ddbName, to make sure we're testing that use case.

File: src/main/resources/ColumnDefinition.json [hunk 30]
```diff
@@ -0,0 +1,34 @@
+[
+  {
+    "name": "healthCode",
+    "columnType": "STRING",
```
Comment (DwayneJengSage) [inline src/main/resources/ColumnDefinition.json]: Similarly, remove columnType from this file.

File: src/main/java/org/sagebionetworks/bridge/exporter/synapse/TransferMethod.java [hunk 31]
```diff
@@ -0,0 +1,66 @@
+package org.sagebionetworks.bridge.exporter.synapse;
+
+import com.amazonaws.services.dynamodbv2.document.Item;
+import com.amazonaws.services.s3.transfer.Transfer;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/synapse/TransferMethod.java]: Minor: s3.transfer.Transfer is unused.

File: src/test/java/org/sagebionetworks/bridge/exporter/handler/HealthDataExportHandlerTest.java [hunk 32]
```diff
@@ -43,6 +44,15 @@
 import org.sagebionetworks.bridge.rest.model.UploadFieldType;
 
 public class HealthDataExportHandlerTest {
+    private static List<ColumnModel> MOCK_COLUMN_LIST;
+
+    private static List<ColumnDefinition> MOCK_COLUMN_DEFINITION;
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/handler/HealthDataExportHandlerTest.java]: Mark these as final. Same with the other 3 test files.

File: src/test/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandlerTest.java [hunk 33]
```diff
@@ -461,13 +476,95 @@ public static void validateTsvHeaders(String line, String... extraColumnNameVara
...
+    }
+
+    public static List<ColumnDefinition> createTestSynapseColumnDefinitions() {
+        // setup column definitions
+        ImmutableList.Builder<ColumnDefinition> columnDefinitionsBuilder = ImmutableList.builder();
+
+        ColumnDefinition healthCodeDefinition = new ColumnDefinition();
+        healthCodeDefinition.setName("healthCode");
+        healthCodeDefinition.setDdbName("healthCode");
+        healthCodeDefinition.setColumnType(ColumnType.STRING);
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandlerTest.java]: ddbName and columnType not needed here. (Similarly in the next 4 columns.)

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 34]
```diff
@@ -202,4 +217,123 @@ public void shouldConvertFreeformTextToAttachment() {
...
+        final String testStringName = "test_string";
+        final String testStringSetName = "test_string_set";
+        final String testDateName = "test_date";
+        final String testExporterDateName = "test_exporter_date";
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: Minor: Exporter date doesn't come from DDB record. We should remove this.

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 35]
```diff
@@ -202,4 +217,123 @@ public void shouldConvertFreeformTextToAttachment() {
...
+        // setup mock
+        ExportTask mockTask = mock(ExportTask.class);
+        when(mockTask.getExporterDate()).thenReturn(TEST_LOCAL_DATE);
+
+        // create mock record
+        Item testRecord = new Item();
+        testRecord.withString(testStringName, "test_string_value");
+        testRecord.withStringSet(testStringSetName, new String[]{"test_string_set_value_1", "test_string_set_value_2"});
+        testRecord.withLong(testDateName, 1484181511);
+        testRecord.withString(testExporterDateName, "test_exporter_date_value");
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java]: Also remove this.

File: src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java [hunk 36]
```diff
@@ -202,4 +217,123 @@ public void shouldConvertFreeformTextToAttachment() {
...
+        List<ColumnDefinition> testColumnDefinitions;
+
+        ImmutableList.Builder<ColumnDefinition> columnDefinitionBuilder = ImmutableList.builder();
+
+        ColumnDefinition testDefinition1 = new ColumnDefinition();
+        testDefinition1.setName(testStringName);
+        testDefinition1.setColumnType(ColumnType.STRING);
+        testDefinition1.setMaximumSize(36L);
+        testDefinition1.setTransferMethod(TransferMethod.STRING);
+        testDefinition1.setDdbName(testStringName);
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/exporter/util/BridgeExporterUtilTest.java:295]: Similarly, remove columnType from here and from the other 3 columns.

File: src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java [hunk 37]
```diff
@@ -47,8 +46,13 @@
     private static final Logger LOG = LoggerFactory.getLogger(SynapseExportHandler.class);
 
     // Package-scoped to be available to unit tests.
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java]: Minor: Remove obsolete comment.

File: src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java [hunk 38]
```diff
@@ -47,8 +46,13 @@
...
-    static final List<ColumnModel> COMMON_COLUMN_LIST;
-    static {
+    private List<ColumnModel> commonColumnList;
+
+    private List<ColumnDefinition> columnDefinition;
+
+    private void setSynapseColumnDefinitionsAndColumnList() {
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/exporter/handler/SynapseExportHandler.java]: Minor: Rename this to something like "initSynapseColumnDefinitionsAndColumnList", so we don't get it confused with an actual setter.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

**Review by DwayneJengSage — APPROVED** [review]

**Comment by DwayneJengSage** [comment]
Looks good. We just need to resolve conflicts before we can merge.
